### DISCUSSION TITLE
Can we add a realmRole at the time of user creation using admin APIs

### DISCUSSION THREAD

**Discussion by Venkat-Siva** [keycloak/keycloak#22698] [Q&A]
Hi all,

I am trying to create a user using admin API (POST /{realm}/users) using following body:
{
    "username": "testUser1",
    "groups" : ["testGrp1"],
    "realmRoles" : ["testRole1"],
    "enabled": true,
    "totp": false,
    "emailVerified": false,
    "credentials": [
        {
            "type": "password",
            "value": "xxxxxx"
        }
    ]
}

I got a 201 response and the user (testUser1) was created successfully. The user is added to the 'testGrp1' group but not mapped to realmRole 'testRole1'.

From other forums (https://stackoverflow.com/questions/73721142/when-creating-a-new-user-set-realmroles-keycloak-admin-rest-api) I see that realm Role can't be added at the time of user creation and it has to be done separately using role mapping API ( POST /{realm}/users/{id}/role-mappings/realm). But as per Keycloak API documentation, there is an optional field for realmRole which we can use to assign roles during user creation.

Can anyone share your experience?

**Comment by christian-2** [comment]
I'm adding a role (`view-users` from client `realm-management` to a special user that represents another client's service account) as follows:

```
kcadm.sh add-roles --config $config -r $realm \
        --uusername service-account-$client_id \
        --cclientid realm-management \
        --rolename view-users
```

So this is an instance where adding an (albeit client-level, not realm-level) role does indeed work when done in a separate step (REST API call) from a user's creation.

**Reply by Venkat-Siva** [reply]
I understand that role can be assigned in a separate step but my question is to check if we can create a user and add a realm role to the user in one API call (at the time of user creation) rather than a separate API call to assign a realmRole to a user. I would like to know the steps if you have managed to do this. Thanks.
